gemd/units/impl.py

```diff
@@ -3,70 +3,132 @@
+            )
+
     global _REGISTRY
-    tokens = list(tokenizer(input_string))
-    scales = []
-
-    unrecognized = [t for t in tokens if t.type == OP and t.string not in _ALLOWED_OPERATORS]
-    if len(unrecognized) > 0:
-        raise UndefinedUnitError(f"Unrecognized operator(s): {unrecognized}")
-
-    # Ignore leading numbers & operators, since Pint handles those itself
-    start = next((i for i, token in enumerate(tokens) if token.type == NAME), len(tokens))
-
-    for i, token in enumerate(tokens[start:], start=start):
-        if token.type != NUMBER:
-            continue
-
-        # Note we can't run off the front because we started at a NAME
-        first = i
-        while tokens[first - 1].string in {'+', '-'}:
-            first -= 1  # Include unary operations
-
-        if tokens[first - 1].string in {"^", "**"}:
-            continue  # Don't mangle exponents
-
-        # Names couple tightly to their preceding numbers, so is it a denominator?
-        division = tokens[first - 1].string in {"/", "//"}
-        tight = i < len(tokens) - 2 and tokens[i + 1].type == NAME
-
-        # Get the number
-        substr = input_string[tokens[first].start[1]:token.end[1]]
-        value = eval(substr)
-        if value <= 0:
-            raise DefinitionSyntaxError(f"Scaling factors must be positive: {substr}")
-        scales.append([substr, token.string, division and tight])
-
-    for substr, value, division in scales:
-        # There's probably something to be said for stashing these, but this sin
-        # should be ameliorated by the LRU cache
-        regex = rf"(?<!=[-+0-9.]){re.escape(substr)}(?!=[0-9.])"
-        valid = "_" + value.replace(".", "_").replace("+", "").replace("-", "_")
-        trailing = "/" if division else ""
-        _REGISTRY.define(f"{valid} = {value} = {value}")
-        input_string = re.sub(regex, valid + trailing, input_string)
+    for scaled_term, number_string, unit_string in todo:
+        regex = rf"(?<![-+0-9.]){re.escape(scaled_term)}(?![0-9.])"
+        stripped = re.sub(r"--", "", re.sub(r"[+\s]+", "", scaled_term))
+
+        if unit_string is not None:
+            stripped_unit = re.sub(r"--", "", re.sub(r"[+\s]+", "", unit_string))
+            long_unit = f"{_REGISTRY(stripped_unit).u}"
+            short_unit = f"{_REGISTRY(stripped_unit).u:~}"
+            long = stripped.replace(stripped_unit, "_" + long_unit)
+            short = stripped.replace(stripped_unit, " " + short_unit)
+        else:
+            long = stripped
+            short = stripped
+
+        underscored = re.sub(r"[-.]", "_", long)
+        valid = f"_{underscored}"
+        if valid not in _REGISTRY:
+            # Parse subexpression to clean things up for define
+            value = f"{_REGISTRY.parse_expression(scaled_term)}"
+            _REGISTRY.define(f"{valid} = {value} = {short}")
+        input_string = re.sub(regex, valid, input_string)
```

Comment:
This normalizes the scaling factor expressions, and if they are novel, stores them in the unit registry.